torch/csrc/jit/python/pybind_utils.h

```diff
@@ -948,10 +950,71 @@ inline py::object runAndInsertCall(
   return toPyObject(std::move(stack.back()));
 }
 
+inline c10::optional<py::object> maybeTorchFunctionDispatch(
+    const py::object& callee,
+    const tuple_slice& args_no_self,
+    const py::kwargs& kwargs,
+    const c10::QualifiedName qualname) {
+  std::vector<py::handle> args_vec = {callee};
+  for (const auto& arg : args_no_self) {
+    args_vec.push_back(arg);
+  }
+  py::tuple args = py::cast(args_vec);
+
+  // Handle __torch_function__ dispatch
+  std::vector<py::handle> overloaded_args;
+  size_t total_arg_num = args.size() + kwargs.size();
+  for (size_t i = 0; i < args.size(); ++i) {
+    is_tensor_and_append_overloaded(args[i].ptr(), &overloaded_args);
+    is_tensor_list_and_append_overloaded(
+        args[i].ptr(),
+        &overloaded_args,
+        static_cast<int>(total_arg_num),
+        false /* throw_error */);
+  }
+  // NB: for kwargs, we cannot guarantee the order of appending
+  // is the same as the argument order in operator's schema.
+  // This is suboptimal, but should be fine. Later when we have
+  // better schema matching and argument parsing, we could
+  // match the operator in `operations` first, then the order will
+  // be guaranteed.
+  for (auto item : kwargs) {
+    is_tensor_and_append_overloaded(item.second.ptr(), &overloaded_args);
+    is_tensor_list_and_append_overloaded(
+        item.second.ptr(),
+        &overloaded_args,
+        total_arg_num,
+        false /* throw_error */);
+  }
+  if (overloaded_args.size() > 0) {
+    std::vector<py::object> overloaded_types;
+    overloaded_types.reserve(overloaded_args.size());
+    for (auto& oarg : overloaded_args) {
+      overloaded_types.push_back(
+          py::reinterpret_borrow<py::object>((PyObject*)Py_TYPE(oarg.ptr())));
+    }
+    py::tuple py_types = py::cast(overloaded_types);
+
+    return pybind11::reinterpret_steal<py::object>(
+        handle_torch_function_no_python_arg_parser(
+            overloaded_args,
+            args.ptr(),
+            kwargs.ptr(),
+            qualname.name().c_str(),
+            args[0].ptr(),
+            qualname.prefix().c_str()));
+  }
```

Comment:
@hameerabbasi To be clear, this function is a direct copy-paste from the implementation for custom ops: https://github.com/pytorch/pytorch/blob/1d6bd157902d4b1347a5d03122d02b407658e263/torch/csrc/jit/python/init.cpp#L1020

So any deficiencies you're pointing out also apply there and should probably be split off into issues, as this points out deficiencies in the current implementation of `__torch_function__`.

Reply:
Code replication is usually bad -- We should think about extracting this somewhere so there aren't 4+ implementations to change when we fix these issues:

- Main C++ one
- This one
- The one you refer to in your comment
- The Python one
- Did I miss any? 🙂 

In any case, we can split it out into issues; but we should deduplicate first.

Reply:
@hameerabbasi If it's okay, I'd prefer to land something like this PR before deduplicating. Lack of support for these methods is blocking some important use cases now. IMO landing duplicated code now should be OK to unblock so long as we make sure that someone goes in to deduplicate afterward, and make sure nobody's diverging the implementations in the intervening time. Does that sound good to you?

Reply:
It sounds good, I’ll open an issue for that tomorrow.

Reply:
@hameerabbasi What James said above; we actually really need this functionality internally for enabling some of our models to be FX+TorchScript compatible. It would be great if we could have this landed this soon and then revisit the refactoring in a later PR.

Reply:
I won't block this PR based on that, I'll comment here once the issue is open.